ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/manage/remote/FileSystemRepoManager.java

```diff
@@ -0,0 +1,133 @@
+            long start = System.currentTimeMillis();
+            downloadFile(latestCompressedGraphInRepoPath, downloadedCompressedGraphFile);
+
+            long end = System.currentTimeMillis();
+            LOGGER.info("[%s] Download of compressed graph file finished after %d ms".formatted(getProfileDescriptiveName(), end - start));
+        } catch (Exception e) {
+            LOGGER.error("[%s] Caught an exception during graph download check or graph download:".formatted(getProfileDescriptiveName()), e);
+        }
+    }
+
+    GraphInfo downloadLatestGraphInfoFromRepository() {
+        GraphInfo latestGraphInfoInRepo = new GraphInfo();
+        LOGGER.debug("[%s] Checking latest graphInfo in remote repository...".formatted(getProfileDescriptiveName()));
+
+        Path latestGraphInfoInRepoPath = Path.of(graphsRepoPath, graphsRepoName, graphsProfileGroup, graphsRepoCoverage, graphsRepoGraphVersion, orsGraphRepoStrategy.getRepoGraphInfoFileName());
+        if (!latestGraphInfoInRepoPath.toFile().exists()) {
+            LOGGER.info("[%s] No graphInfo found in remote repository: %s".formatted(getProfileDescriptiveName(), latestGraphInfoInRepoPath.toFile().getAbsolutePath()));
+            return latestGraphInfoInRepo;
+        }
+
+        File downloadedGraphInfoFile = orsGraphFileManager.getDownloadedGraphInfoFile();
+        downloadFile(latestGraphInfoInRepoPath, downloadedGraphInfoFile);
+
+        if (downloadedGraphInfoFile.exists()) {
+            Path latestCompressedGraphInRepoPath = Path.of(graphsRepoPath, graphsRepoName, graphsProfileGroup, graphsRepoCoverage, graphsRepoGraphVersion, orsGraphRepoStrategy.getRepoCompressedGraphFileName());
+            URI uri = latestCompressedGraphInRepoPath.toUri();
+            latestGraphInfoInRepo.withRemoteUri(uri);
+
+            PersistedGraphInfo persistedGraphInfo = orsGraphFileManager.readOrsGraphInfo(downloadedGraphInfoFile);
+            latestGraphInfoInRepo.withPersistedInfo(persistedGraphInfo);
+        } else {
+            LOGGER.error("[%s] Invalid download path for graphInfo file: %s".formatted(getProfileDescriptiveName(), latestGraphInfoInRepoPath));
+        }
+
+        return latestGraphInfoInRepo;
+    }
+
+    public void downloadFile(Path repoPath, File localPath) {
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("[%s] Downloading %s to local file %s...".formatted(getProfileDescriptiveName(), repoPath.toFile().getAbsolutePath(), localPath.getAbsolutePath()));
+        } else {
+            LOGGER.info("[%s] Downloading %s...".formatted(getProfileDescriptiveName(), repoPath.toFile().getName()));
+        }
+        if (repoPath != null) {
+            try {
+                Files.copy(repoPath, localPath.toPath(), StandardCopyOption.REPLACE_EXISTING);
+            } catch (IOException e) {
+                LOGGER.warn("[%s] Caught %s when trying to download %s".formatted(getProfileDescriptiveName(), repoPath.toFile().getAbsolutePath()));
```

Comment:
The format string contains three parameters %s but only two are used. Might result in undefined behavior.

Reply:
Fixed

Reply:
Where?